test/integration/http_server_options_test.lua

```diff
@@ -87,3 +87,80 @@ g.test_disable_keepalive_default = function(g)
     local httpd = g.httpd
     t.assert_equals(table.getn(httpd.options.disable_keepalive), 0)
 end
+
+g.before_test('test_idle_timeout_default', function()
+    g.httpd = helpers.cfgserv()
+    g.httpd:start()
+end)
+
+g.test_idle_timeout_default = function(g)
+    local httpd = g.httpd
+    t.assert_equals(httpd.options.idle_timeout, 0)
+end
+
+g.before_test('test_idle_timeout_is_set', function()
+    g.idle_timeout = 0.5
+    g.httpd = helpers.cfgserv({
+        idle_timeout = g.idle_timeout,
+    })
+    g.httpd:start()
+end)
+
+g.test_idle_timeout_is_set = function(g)
+    local conn_is_opened = false
+    local conn_is_closed = false
+    local httpd = g.httpd
+    g.httpd.internal.preprocess_client_handler = function() conn_is_opened = true end
+    g.httpd.internal.postprocess_client_handler = function() conn_is_closed = true end
+
+    t.assert_equals(httpd.options.idle_timeout, g.idle_timeout)
+
+    -- Set HTTP keepalive headers: Connection:Keep-Alive and
+    -- Keep-Alive:timeout=<keepalive_idle>. Otherwise HTTP client will send
+    -- "Connection:close".
+    local opts = {
+        keepalive_idle = 3600,
+        keepalive_interval = 3600,
+    }
+    local r = http_client.new():request('GET', helpers.base_uri .. '/test', nil, opts)
+    t.assert_equals(r.status, 200)
+    t.assert_equals(r.headers.connection, 'keep-alive')
+    t.assert_equals(conn_is_opened, true)
+
+    helpers.retrying({
+        timeout = g.idle_timeout * 2,
+    }, function()
+        if conn_is_closed ~= true then
+            assert()
+        end
+    end)
+    t.assert_equals(conn_is_closed, true) -- Connection is closed.
+end
+
+g.before_test('test_idle_timeout_is_unset', function()
+    g.httpd = helpers.cfgserv({})
+    g.httpd:start()
+end)
+
+g.test_idle_timeout_is_unset = function(g)
+    local conn_is_opened = false
+    local conn_is_closed = false
+    g.httpd.internal.preprocess_client_handler = function() conn_is_opened = true end
+    g.httpd.internal.postprocess_client_handler = function() conn_is_closed = true end
+
+    -- Set HTTP keepalive headers: Connection:Keep-Alive and
+    -- Keep-Alive:timeout=<keepalive_idle>. Otherwise HTTP client will send
+    -- "Connection:close".
+    local opts = {
+        keepalive_idle = 3600,
+        keepalive_interval = 3600,
+    }
+    local r = http_client.new():request('GET', helpers.base_uri .. '/test', nil, opts)
+    t.assert_equals(r.status, 200)
+    t.assert_equals(r.headers.connection, 'keep-alive')
+    t.assert_equals(conn_is_opened, true)
+    helpers.retrying({}, function()
+        return conn_is_closed == false
+    end)
```

Comment:
The same as for `helpers.retrying()` in `helpers.teardown()`: we should *raise* an error when we don't meet an expected condition.

```diff
@@ -160,7 +158,7 @@ g.test_idle_timeout_is_unset = function(g)
     t.assert_equals(r.headers.connection, 'keep-alive')
     t.assert_equals(conn_is_opened, true)
     helpers.retrying({}, function()
-        return conn_is_closed == false
+        assert(conn_is_closed == false, 'assert that the connection is closed')
     end)
     t.assert_equals(conn_is_closed, false) -- Connection is alive.
 end
```

Reply:
Looked a bit deeper into the test case. `conn_is_closed` is `false` initially, there is no sense to wait for this situation: `helpers.retrying()` will always return immediately. We can either check that the connection was not closed by the server immediately or wait for some short time and check.

For example:

```diff
@@ -159,8 +157,5 @@ g.test_idle_timeout_is_unset = function(g)
     t.assert_equals(r.status, 200)
     t.assert_equals(r.headers.connection, 'keep-alive')
     t.assert_equals(conn_is_opened, true)
-    helpers.retrying({}, function()
-        return conn_is_closed == false
-    end)
     t.assert_equals(conn_is_closed, false) -- Connection is alive.
 end
```

Or:

```diff
diff --git a/test/integration/http_server_options_test.lua b/test/integration/http_server_options_test.lua
index 504891d..ef2954f 100644
--- a/test/integration/http_server_options_test.lua
+++ b/test/integration/http_server_options_test.lua
@@ -1,3 +1,4 @@
+local fiber = require('fiber')
 local t = require('luatest')
 local http_client = require('http.client')
 

@@ -158,9 +157,12 @@ g.test_idle_timeout_is_unset = function(g)
     local r = http_client.new():request('GET', helpers.base_uri .. '/test', nil, opts)
     t.assert_equals(r.status, 200)
     t.assert_equals(r.headers.connection, 'keep-alive')
+
+    -- The server should not close the connection during the
+    -- keepalive_idle seconds. We can't wait so long in the test,
+    -- so let's wait for a short time and check.
+    fiber.sleep(0.5)
+
     t.assert_equals(conn_is_opened, true)
-    helpers.retrying({}, function()
-        return conn_is_closed == false
-    end)
     t.assert_equals(conn_is_closed, false) -- Connection is alive.
 end
```